compiler/noirc_evaluator/src/ssa/opt/bubble_up_constrains.rs

```diff
@@ -20,19 +30,36 @@ impl Ssa {
                         }
                     };
 
-                    let index = filtered_instructions
+                    let last_instruction_that_creates_inputs = filtered_instructions
                         .iter()
                         .rev()
-                        .position(|instruction_id| {
-                            let results = dfg.instruction_results(*instruction_id).to_vec();
+                        .find(|&&instruction_id| {
+                            let results = dfg.instruction_results(instruction_id).to_vec();
                             results.contains(&lhs) || results.contains(&rhs)
                         })
-                        // We iterate through the previous instructions in reverse order so the index is from the
-                        // back of the vector. Subtract from vector length to get correct index.
-                        .map(|reversed_index| filtered_instructions.len() - reversed_index)
-                        .unwrap_or(0);
+                        .copied();
+
+                    let insertion_index = last_instruction_that_creates_inputs
+                        .map(|id| {
+                            filtered_instructions
+                                .iter()
+                                .position(|&x| x == id)
+                                .expect("Instruction should exist")
+                                // We want to insert just after the last instruction that creates the inputs
+                                + 1
+                        })
+                        .unwrap_or_default();
```

Comment:
I didn't want to do it that way because the position is the reversed one, so I felt that it was clearer this way. Let me see if I can avoid the double search but try to keep the code as readable as possible

Reply:
We immediately fix the reversed index however so shouldn't it be ok? We may need an off-by-one adjustment though.

Reply:
Latest commit looks good.